### ISSUE TITLE
Feature request: shadow DOM support

### ISSUE THREAD

**Issue by kr05** [Simonwep/pickr#86]
#### Do you want to request a **feature** or report a **bug**? 
I'm not sure if it's a feature request or if it's actually meant to work with the shadow DOM, but Pickr currently has trouble working inside the shadow DOM.

#### What is the current behavior?
Named/default imports don't work. I kept getting the following error: 
![image](https://user-images.githubusercontent.com/11833744/57170523-462dcb00-6dc2-11e9-974e-c201212aa904.png)

To fix it, I had to import the files as follows:
```javascript
import '@simonwep/pickr/dist/pickr.min.js'; 
//Workaround for styles not being applied completely
import { PickrDefaultStyles } from './pickr-default-styles';
```
Getting the styles to work wasn't too much of a hassle. I use lit-element as my component library, which provides a handy CSS utility. I simply create a new file and wrap the CSS inside the following:
```javascript
import { css } from 'lit-element';
export const PickrDefaultStyles = css`
     //css goes here
`;
```
Although the files load correctly, Pickr.create doesn't work with selectors. You have to use actual DOM elements. No big deal, but this leads to the final error. Pickr crashes when calling replaceChild. Apparently parentElement doesn't exist or isn't accessible. Here's the error:
![image](https://user-images.githubusercontent.com/11833744/57168591-3e1c5e00-6db7-11e9-8d53-4177bd017eba.png)

Finally, I replaced all instances of parentElement with parentNode and it works as expected. Kind of. Unfortunately, it doesn't respond to clicks or taps. Loading the css through the CDN in my index file makes it work. If I remove the mentioned CSS import, the pickr-button isn't styled correctly, so I have to load both instances of the stylesheet, once in my index file and the second inside the component. 

In conclusion, to make it work I had to:

1. Import pickr without named/default imports
2. Import CSS through both CDN in my index file and inside component.
3. Use an actual DOM element in Pickr.create, not a selector.
4. Replace parentElement with parentNode inside the pickr js file.

Here's the [JSFiddle](https://jsfiddle.net/8afc35hL/) containing the closest representation of my environment. I'm simulating the CSS import by creating a constant inside the component and treating that as if it had been imported. I also left the original Pickr element to demonstrate that the files have indeed been loaded.

```
Pickr-version (see Pickr.version): 0.4.11
Used Pickr-bundle (es5 or normal one): normal one
Browser-version:  Chrome 73.0.3683.103
Operating-system:  Windows 10
```

**Comment by Simonwep** [comment]
Wow, first of all - thanks for this extremely detailed description and fiddle!

Unfortunately, no. I never took the Shadow DOM into consideration, but now I will.
I'll work on providing a default export (if possible, never tried it before) and Shadow DOM support :)

**Comment by Simonwep** [comment]
Okay, the problem with `parentNode` was easily solved. Import with `import Pickr from '@simonwep/pickr'` or `import Pickr from '[node-modules-path]'`  should work, just tested it. The usage of selectors without having a reference to the shadow root is not possible? The usage of providing a direct reference would be indispensable 😕

**Comment by kr05** [comment]
I'm glad you found it useful, it's the least I could when you've provided such an amazing library 👍

It wasn't working before the changes the parentElement. I don't remember testing it again after the changes, maybe that gets fixed as well! As for the parentNode problem, I'm glad it wasn't too difficult.

**Comment by kr05** [comment]
![image](https://user-images.githubusercontent.com/11833744/57184020-ea1e8180-6e69-11e9-9bd3-dcaac0b4bea6.png)

This is the error that I get when I use a reference, such as an id or class name.

**Comment by Simonwep** [comment]
As far as I know you can't use a selector as reference to a element I a shadow DOM, or in which env does this happen? :suspect:

**Comment by kr05** [comment]
That's an interesting point, I hasn't considered it. The shadow DOM provides some difficulties in that regard. I've seen other libraries struggle with the issue of not being able to use selectors because the actual element is not globally available. Maybe if the intialization of Pickr were constrained within the component that actually imports it, it would simply be a matter of using 

```javascript
this.shadowRoot.querySelector(selector)
```

I haven't checked how Pickr uses the given selectors, so I can't provide more insight. My environment is the browser, using lit-element as my component library. Pretty bare bones, so it's mainly the browser doing the heavy lifting.

**Comment by Simonwep** [comment]
Okay I have a solution to this shadow root problem. Consider this DOM:
```html
<div class="entry">
  #shadow-root 
    <div class="innr">
       <div class="another">
         #shadow-root 
           <div class="pickr"></div>
       </div>
    </div>
</div>
```
This could be accessed using a custom `>>>` selector. Eg.
```js
{
  el: '.entry >>> .innr .another >>> .pickr' 
}
```

Pickr would resolve the shadow root of each element before an `>>>` with `aelement.shadowRoot`.
Would this be a good solution?

**Comment by kr05** [comment]
It seems a bit unwieldy, especially if the component is nested deep. 
Have you tried doing a null check for this.shadowRoot? If it isn't null, use opt.el with this.shadowRoot.querySelector, otherwise use the document.

**Comment by Simonwep** [comment]
Yeah but to what refers `this.shadowRoot`? That would be a direct reaction to your individual environment, and not a general / universal solution... or what do you mean? using `>>` or `>>>` would allow deep nesting, more options would be to recursive travel the tree and search for the selector or providing the element itself.

**Comment by kr05** [comment]
I have an alternative solution, lmk what you think! Instead of having shadow DOM users provide the full "path" to their element, they can just provide their context.

```javascript
//providing my context to the factory method
ctx: this,
el: "#pickrSelector"
```

In the constructor, check if ctx exists. If it does, assign it to the opt object.

```javascript
function constructor(opt) {

   //Previous lines unchanged...

   //Extract context
   const {swatches, inline, components, ctx} = opt;

   //Middle lines unchanged...

   // Assign ctx if provided 
   if (ctx) {
      opt.ctx = ctx;
   }
}
```

```javascript
function _preBuild() {

   //Previous lines unchanged...

   let root = opt.ctx ? opt.ctx.shadowRoot : document

   // Check if element is selector
   if (typeof opt.el === 'string') {
      opt.el = root.querySelector(opt.el);
   }
}
```

Finally, in the _preBuild function, check for the ctx property again. If it exists, call querySelector using the context, otherwise continue using document.

There are multiple flavors of this solution, but it's hard to test out all the variations without knowing your dev environment and being able to change the pickr.js file. I successfully tested a simpler version of this solution by modifying the minified file, which was...a first for me, and quite interesting. The most promising variation is the following:

- Have a global ctx variable that gets assigned in the preBuild step. Replace every document call with this new ctx variable.

This would make it simple to make changes to the context, without requiring making changes the every function that required document. Hope this helps!

**Comment by Simonwep** [comment]
I generally want a rather easy and universal config which could be adapted into any env. Defining a ctx would assume to describe a new option, moreover it would only provide shadowRoot so it would be kinda redundant since you could directly provide `shadowRoot` into pickr's config.

And I think if you already have the shadowRoot as object, you can also directly provide the element which should be pickr as `el.`.

Introducing `>>` or `>>>` as selector would make a additional option redundant without losing the possibility to select your element with a css selector.  A `ctx` would IMO make things unnecessarily complicated. Or do you have another, equally simple solution?

Your idea is good, no question. But really unpractical for the targets and principles of pickr 😕

**Comment by kr05** [comment]
I don't, unfortunately...that's the only solution that came to mind. At the end of the day the user doesn't care about the details of implementation, so I'm excited to see Pickr support shadowRoot regardless of the solution you decide to use!

**Comment by Simonwep** [comment]
Shadow-dom selector are available since 0.5.0, If there are any features which could be added - just comment and I'll reopen this issue :)
